Added decompression transformer

blanu opened the pull request:
Added a transformer that uses arithmetic coding for entropy reduction via decompression
Added decompression transformer to protean
Merged uniform probability distribution code that was shared between the transformer and the tests

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-lib/296)

<!-- Reviewable:end -->

Commit:
Added a transformer that uses arithmetic coding for entropy reduction

Commit:
Enabled decompression shaping

Commit:
Deduplicated sample config code with decompressionShaper

trevj commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 35 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzjQf-KKT_GdpfCN5Y)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L35)):</sup>
Can you have the constructor call `configure`? That way, there's no need to specify a config when testing with Docker. (I will clean up this mess at some future date)

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzjfT6K8DVfLLBbOv3)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L45)):</sup>
` = `!

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 59 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkaTWblCClXhu7Jd_)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L59)):</sup>
What is this magic value? I think extracting a named constant would help readability.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkzPxS-Vz4w5ZngXW)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L60)):</sup>
What does the footer do? Can you describe the format of the packets sent over the wire?

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 68 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkpX2_7S4XZ98JVrF)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L68)):</sup>
Intuitively, I would expect `restore` to use a *decoder*. Did you deliberately choose this naming scheme?

---

<sup>**[src/fancy-transformers/protean.ts, line 102 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzlCg6RspBZnfA2Y2s)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/protean.ts#L102)):</sup>
I'm not a big fan of commented code...I'm sure this was useful during testing but always better to have some unit tests instead, which you have.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296)
<!-- Sent from Reviewable.io -->

Commit:
Removed debug logging

Commit:
Use const instead of var for logger

Commit:
Use function syntax instead of var syntax for exported functions

Commit:
Use let instead of var throughout

Commit:
Added comments on why the compresser is being used backwards.

Commit:
Added comments on arithmetic coder encoding format

Commit:
Use spaces around =

Commit:
Have the constructor call configure

Commit:
Removed syntactically incorrect => in function declaration

blanu commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 35 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzjQf-KKT_GdpfCN5Y)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L35)):</sup>
Ah, I missed it when you did this to the other transformers. I've made this change now.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 45 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzjfT6K8DVfLLBbOv3)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L45)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 59 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkaTWblCClXhu7Jd_)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L59)):</sup>
I have added comments about this.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkzPxS-Vz4w5ZngXW)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L60)):</sup>
I have added comments about this.

---

<sup>**[src/fancy-transformers/decompressionShaper.ts, line 68 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzkpX2_7S4XZ98JVrF)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/decompressionShaper.ts#L68)):</sup>
Yes! Everything is backwards from what you'd normally expect because this transformer in fact works backwards from what you'd normally expect. I have added comments to this effect.

---

<sup>**[src/fancy-transformers/protean.ts, line 102 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296#-JzzlCg6RspBZnfA2Y2s)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/80e955e1363a59f056721053786aefddb5418f60/src/fancy-transformers/protean.ts#L102)):</sup>
Oops, I meant to remove these. I've made this change now.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296)
<!-- Sent from Reviewable.io -->

Commit:
Fixed a bug where the decompressor was truncating output

trevj commented:
:+1:

Yay!

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/296)
<!-- Sent from Reviewable.io -->
